docs/how-to/push-update-on-mainnet.md

```diff
@@ -0,0 +1,192 @@
+  * https://cardanodocs.com/technical/updater/
+1. Cluster nodes' keys. They are crucial to have since cluster nodes
+   are the only nodes that have stake in the system so they are the
+   only nodes that can propose an update and vote for it. It's enough
+   to have keys for majority of stake. In mainnet we have 7 core nodes
+   with equal stake, so 4 keys should be enough.
+2. Access to the S3 bucket
+3. Software version
+   - Single integer, must be greater by 1 than the last confirmed
+     version. Initially `csl-daedalus` application has version 0. The
+     first update should have version 1, the next one 2, etc. If you
+     try to propose an update with version 3 while the last confirmed
+     one is 1, your update proposal won't be accepted by the nodes.
+   - Node knows its software version from `configuration.yaml` file,
+     `applicationVersion` field. When you propose an update to version
+     X, make sure that all attached installers use configuration with
+     `applicationVersion` set to `X`.
+4. Installers corresponding to software version we're pushing to cluster
+   - Those are two executables (exe for win, pkg for mac) which are to
+     be provided by QA (after QA procedures passed). They are
+     installers provided by our CI.
+   - Advice: after providing installers, ask QA to confirm hashes are
+     same, we don't want to ship users incorrect installers to users.
+   - IMPORTANT: pushing an installer of version N with an update with version 
+     M > N will result in the infinite installer loop (Daedalus thinks 
+     it is N, it downloads M, updates to it, but N is installed repeats), so
+     versions should be double-checked. 
+
+
+Proposing the update
+====================
+
+Preparations
+------------
+
+You need to build packages `cardano-sl-tools` and `cardano-sl-auxx`. Branch's code should be compatible with cluster (take release branch, e.g. `cardano-sl-1.0`).
+
+Ensure your keys are somewhere around and named, say, `keyX.sk`, where `X ∈ [0..6]`.
+
+Before running all commands below, set command-line options appropriately. You need to change `CONFIG_KEY`, set `RELAY_PEER` (a host of any relay node) and possibly setting `--log-config`, `--logs-prefix` to real paths. Note that `--system-start 0` is perfectly valid and you don't need to change it.
+
+The following variables will be used in all following commands:
+
+```
+CONFIG_KEY=mainnet_dryrun_full  # Should be the same as on the running nodes!
+RELAY_PEER=<IP>:3000            # A host of any relay node.
+
+COMMONOPTS="--system-start 0 --configuration-file ../cardano-sl/node/configuration.yaml --configuration-key ${CONFIG_KEY}"
+
+AUXXOPTS="--log-config ../cardano-sl/scripts/log-templates/log-config-qa.yaml --logs-prefix logs/aux-update.1.0.1 --db-path aux-update-1.0.1 --peer ${RELAY_PEER}"
+```
+
+Rearranging the keys
+--------------------
+
+Apply `cardano-keygen rearrange` to the keys to make them usable for proposing an update and voting:
```

Comment:
Shouldn't rearrange be ran after importing? If not, then I don't understand what it does.

Reply:
No, rearrangement should be executed before importing. In fact, key won't be imported unless it's rearranged. 

User secret has these fields:
```
    { _usKeys      :: [EncryptedSecretKey]
    , _usPrimKey   :: Maybe SecretKey
    ... }
```
And by default `_usKeys` is empty. Rearrangement creates an encrypted key from primary key and puts it into `_usKeys`.

Reply:
@domenkozar ^